fusion/fusion/FieldExtractor.hpp

```diff
@@ -0,0 +1,279 @@
+               const FieldName f1, const OptArg f2 = std::nullopt) const
+    {
+        return !compare(input[a], input[b], f1, f2, EQ);
+    }
+
+    //! \brief greater than
+    bool gt(const InstPtrListType & input, size_t a, size_t b,
+            const FieldName f1, const OptArg f2 = std::nullopt) const
+    {
+        return compare(input[b], input[a], f1, f2, LT);
+    }
+
+    //! \brief less than or equal
+    bool lteq(const InstPtrListType & input, size_t a, size_t b,
+              const FieldName f1, const OptArg f2 = std::nullopt) const
+    {
+        return !compare(input[b], input[a], f1, f2, LT);
+    }
+
+    //! \brief greater than or equal
+    bool gteq(const InstPtrListType & input, size_t a, size_t b,
+              const FieldName f1, const OptArg f2 = std::nullopt) const
+    {
+        return !compare(input[a], input[b], f1, f2, LT);
+    }
+
+    //! \brief return the integer read ports used by the input fusion group
+    uint32_t getIntRdPorts(const InstPtrListType & input) const
+    {
+        return countPorts(input,
+                          &Instruction<uArchInfo>::getIntSourceRegs);
+    }
+
+    //! \brief return the integer write ports used by the input fusion
+    //! group
+    uint32_t getIntWrPorts(const InstPtrListType & input) const
+    {
+        return countPorts(input, &Instruction<uArchInfo>::getIntDestRegs);
+    }
+
+    //! \brief return the float read ports used by the input fusion group
+    uint32_t getFloatRdPorts(const InstPtrListType & input) const
+    {
+        return countPorts(input,
+                          &Instruction<uArchInfo>::getFloatSourceRegs);
+    }
+
+    //! \brief return the float write ports used by the input fusion group
+    uint32_t getFloatWrPorts(const InstPtrListType & input) const
+    {
+        return countPorts(input,
+                          &Instruction<uArchInfo>::getFloatDestRegs);
+    }
+
+    //! \brief return the vector read ports used by the input fusion group
+    uint32_t getVecRdPorts(const InstPtrListType & input) const
+    {
+        return countPorts(input,
+                          &Instruction<uArchInfo>::getVectorSourceRegs);
+    }
+
+    //! \brief return the vector write ports used by the input fusion group
+    uint32_t getVecWrPorts(const InstPtrListType & input) const
+    {
+        return countPorts(input,
+                          &Instruction<uArchInfo>::getVectorDestRegs);
+    }
+
+  private:
+    //! \brief compare common method
+    bool compare(const InstPtrType LHS, const InstPtrType RHS,
+                 const FieldName f1, const OptArg _f2, FUNC func) const
+    {
+        FieldName f2 = _f2.value_or(f1);
+
+        auto lhs = getField(LHS, f1);
+        auto rhs = getField(RHS, f2);
+
+        switch (func)
+        {
+            case FUNC::LT: { return lhs < rhs;  }
+            case FUNC::EQ: { return lhs == rhs; }
+            default:       { return false;      }
```

Comment:
default case should be assert, as we should never call compare with anything other than LT/EQ. It is a coding error.

Reply:
ok